File: sort_py/sort.py
#!/usr/bin/env python3

import sys
import time
import random


def sort_bubble(sample):
    """ Bubble sort - slower version """

    swapped = True
    comp_count = 0 
    sample_len = len(sample)

    while swapped:
        swapped = False
        i = 0
        while i < sample_len - 1:
            comp_count += 1
            if sample[i + 1] < sample[i]:
                sample[i], sample[i + 1] = sample[i + 1], sample[i]
                swapped = True
            i += 1
    return comp_count


def sort_bubble(sample):
    """ Bubble sort - faster version """
    
    comp_count = 0
    sample_len = len(sample)
    swapped = True

    j = -1
    while swapped:
        swapped = False
        j += 1
        for i in range(1, sample_len - j):
            comp_count += 1
            if sample[i - 1] > sample[i]:
                sample[i - 1], sample[i] = sample[i], sample[i - 1]
                swapped = True

    return comp_count


def sort_selection(sample):
    """ Selection sort - slower version """

    updated = True
    comp_count = 0 
    sample_len = len(sample)

    sorted_len = 0
    while sorted_len < sample_len:
        i = sorted_len
        while i < sample_len:
            if sample[i] < sample[sorted_len]:
                sample[i], sample[sorted_len] = sample[sorted_len], sample[i]
            i += 1
            comp_count += 1
        sorted_len += 1
    return comp_count


def sort_selection(sample):        
    """ Selection sort - faster version """

    comp_count = 0

    for i in range(len(sample)):
        minimum = i
        
        for j in range(i + 1, len(sample)):
            comp_count += 1
            if sample[j] < sample[minimum]:
                minimum = j

        sample[minimum], sample[i] = sample[i], sample[minimum]
            
    return comp_count


def sort_insertion(sample):
    """ Insertion sort - modified (left to right seek) """

    comp_count = 0
    sample_len = len(sample)

    for j in range(sample_len):
        for i in range(j, sample_len):
            comp_count += 1
            if sample[i] < sample[j]:
                sample.insert(j, sample.pop(i))
        
    return comp_count


def _sort_insertion(sample):
    """ Insertion sort - classic (right to left seek) """

    comp_count = 0
    sample_len = len(sample)

    for j in range(sample_len):
        for i in range(sample_len - 1, j, -1):
            comp_count += 1
            if sample[i] < sample[j]:
                sample.insert(j, sample.pop(i))

    return comp_count


def main(argv):
    if len(argv) < 2:
        print("Please specify algorithm [bubble|selection]")
        return -1
    algorithm = argv[1]

    if len(argv) < 3:
        print("Please specify sample file")
        return -2
    with open(argv[2], "rb") as _:
        sample = list(_.read())

    start = time.time()

    if algorithm == "bubble":
        comp_count = sort_bubble(sample)
    elif algorithm == "selection":
        comp_count = sort_selection(sample)
    elif algorithm == "insertion":
        comp_count = sort_insertion(sample)
    else:
        "Please specify algorithm [bubble|selection]"
        return -3

    run_time = (time.time() - start) * 1000

    if len(argv) == 4 and argv[3] == "result":
        print(sample)

    print(f"language='python', {algorithm=}, {run_time=}ms, {comp_count=}")

if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: sort_c/sort.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <time.h>


u_int64_t sort_bubble(u_int8_t* sample, u_int64_t sample_len)
{
    u_int8_t swapped = 1;
    u_int64_t comp_count = 0;

    while(swapped) {
        u_int64_t i = 0;
        swapped = 0;
        while(i < sample_len - 1) {
            if(sample[i + 1] < sample[i]) {
                u_int8_t _ = sample[i];
                sample[i] = sample[i + 1];
                sample[i + 1] = _;
                swapped = 1;
            }
            i++;
            comp_count++;
        }
    }

    return comp_count;
}


u_int64_t sort_selection(u_int8_t* sample, u_int64_t sample_len)
{
    u_int64_t comp_count = 0;
    u_int64_t sorted_len = 0;
    u_int64_t i;

    while(sorted_len < sample_len) {
        i = sorted_len;;
        while(i < sample_len) {
            if(sample[i] < sample[sorted_len]) {
                u_int8_t _ = sample[sorted_len];
                sample[sorted_len] = sample[i];
                sample[i] = _;
            }
            i++;
            comp_count++;
        }
        sorted_len += 1;
    }

    return comp_count;
}


u_int64_t sort_insertion(u_int8_t* sample, u_int64_t sample_len)
{
    u_int64_t comp_count = 0;

    for(int j = 0; j < sample_len; j++) {
        for(int i = j; i < sample_len; i++) {
            comp_count++;
            if(sample[i] < sample[j]) {
                int element = sample[i];
                for(int n = i; n > j; n--) {
                    sample[n] = sample[n - 1];
                }
                sample[j] = element;
            }
        }
    }

    return comp_count;
}


int main(int argc, char *argv[])
{

    if(argc < 2) {
        printf("Please specify algorithm [bubble|selection]\n");
        return -1;
    }

    if(argc < 3) {
        printf("Please specify sample file\n");
        return -2;
    }

    struct stat st;
    stat(argv[2], &st);
    int sample_len = st.st_size;
    int fd = open(argv[2], O_RDONLY);
    u_int8_t* sample = malloc(sample_len);
    int _ = read(fd, sample, sample_len);
    close(fd);

    u_int64_t comp_count;
    clock_t start = clock();

    if(!strcmp("bubble", argv[1])) {
        comp_count = sort_bubble(sample, sample_len);
    }
    else if(!strcmp("selection", argv[1])) {
        comp_count = sort_selection(sample, sample_len);
    }
    else if(!strcmp("insertion", argv[1])) {
        comp_count = sort_insertion(sample, sample_len);
    }
    else {
        printf("Please specify algorithm [bubble|selection]\n");
        return -1;
    }

    double run_time = (((double) (clock() - start)) / CLOCKS_PER_SEC) * 1000;

    if(argc == 4 && !strcmp("result", argv[3])) {
        printf("[");
        for(int i = 0; i < sample_len - 1; i++) {
            printf("%d, ", sample[i]);
        }
        printf("%d]\n", sample[sample_len - 1]);
    }

    printf("language='C', algorithm='%s', run_time=%lfms, comp_count=%ld\n", argv[1], run_time, comp_count);

    free(sample);

    return 0;
}
